### PR TITLE
Bug 1827723: Make default channel optional

### PR DESCRIPTION
**Pull request by Bowenislandsong**
This commit address the bug where a bundle was released before the current default channel was created, the inference fails.

### CODE REVIEW

**Comment by Bowenislandsong** [comment]
the new annotation for prometheus 0.14.0 in the e2e test:
```
annotations:
  operators.operatorframework.io.bundle.channel.default.v1: ""
  operators.operatorframework.io.bundle.channels.v1: stable
  operators.operatorframework.io.bundle.manifests.v1: manifests/
  operators.operatorframework.io.bundle.mediatype.v1: registry+v1
  operators.operatorframework.io.bundle.metadata.v1: metadata/
  operators.operatorframework.io.bundle.package.v1: prometheus
```

**Comment by Bowenislandsong** [comment]
switched channels and default channel situation for 0.15.0 case so that this bundle is added in the middle between 0.14.0 and 0.22.2

File: manifests/prometheus/0.22.2/prometheusoperator.0.22.2.clusterserviceversion.yaml [hunk 1]
```diff
@@ -7,7 +7,7 @@ metadata:
   annotations:
     alm-examples: '[{"apiVersion":"monitoring.coreos.com/v1","kind":"Prometheus","metadata":{"name":"example","labels":{"prometheus":"k8s"}},"spec":{"replicas":2,"version":"v2.3.2","serviceAccountName":"prometheus-k8s","securityContext": {}, "serviceMonitorSelector":{"matchExpressions":[{"key":"k8s-app","operator":"Exists"}]},"ruleSelector":{"matchLabels":{"role":"prometheus-rulefiles","prometheus":"k8s"}},"alerting":{"alertmanagers":[{"namespace":"monitoring","name":"alertmanager-main","port":"web"}]}}},{"apiVersion":"monitoring.coreos.com/v1","kind":"ServiceMonitor","metadata":{"name":"example","labels":{"k8s-app":"prometheus"}},"spec":{"selector":{"matchLabels":{"k8s-app":"prometheus"}},"endpoints":[{"port":"web","interval":"30s"}]}},{"apiVersion":"monitoring.coreos.com/v1","kind":"Alertmanager","metadata":{"name":"alertmanager-main"},"spec":{"replicas":3, "securityContext": {}}}]'
 spec:
-  replaces: prometheusoperator.0.15.0
+  replaces: prometheusoperator.0.14.0
```
Comment (kevinrizza) [inline manifests/prometheus/0.22.2/prometheusoperator.0.22.2.clusterserviceversion.yaml:10]: Why are we reordering these? I would prefer that, if we need manifests for a specific test, that we make a copy of them and place them in a separate test folder rather than modify our default examples.

File: pkg/lib/bundle/generate.go [hunk 2]
```diff
@@ -12,6 +12,7 @@ import (
 	"gopkg.in/yaml.v2"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/kubectl/pkg/util/slice"
```
Comment (kevinrizza) [inline pkg/lib/bundle/generate.go:15]: Is this external dependency really necessary? In general, I think that utils like this which are not included in the standard library and are trivial to write should just be implemented directly.

File: pkg/lib/bundle/generate.go [hunk 3]
```diff
@@ -298,6 +302,10 @@ func ValidateChannelDefault(channels, channelDefault string) (string, error) {
 	var chanErr error
 	channelList := strings.Split(channels, ",")
 
+	if len(channelList) == 0 || slice.ContainsString(channelList, "", nil) {
+		return chanDefault, fmt.Errorf("invalid channels is provied: %s", channels)
```
Comment (kevinrizza) [inline pkg/lib/bundle/generate.go:306]: This error message seems confusing. Could we be more descriptive of what is actually going wrong here? It seems like we are actually checking two things that should have separate error messages:

1. That no channels were specified
2. That the list of channels included an empty string in the list I think?

Also spelling/grammar nit: use `are` instead of `is` for plurals, and `provided` instead of `provied`

File: pkg/lib/bundle/generate_test.go [hunk 4]
```diff
@@ -207,7 +207,7 @@ func TestGenerateDockerfileFunc(t *testing.T) {
 		"LABEL operators.operatorframework.io.bundle.metadata.v1=%s\n"+
 		"LABEL operators.operatorframework.io.bundle.package.v1=test4\n"+
 		"LABEL operators.operatorframework.io.bundle.channels.v1=test5\n"+
-		"LABEL operators.operatorframework.io.bundle.channel.default.v1=test5\n\n"+
+		"LABEL operators.operatorframework.io.bundle.channel.default.v1=\n\n"+
```
Comment (kevinrizza) [inline pkg/lib/bundle/generate_test.go:210]: Rather than coopting existing tests, can we write an explicit test specifically for this edge case?

File: test/e2e/opm_test.go [hunk 5]
```diff
@@ -75,7 +73,7 @@ func buildBundlesWith(containerTool string) error {
 		bundleTag3: bundlePath3,
 	} {
 		if err := inTemporaryBuildContext(func() error {
-			return bundle.BuildFunc(path, "", bundleImage+":"+tag, containerTool, packageName, channels, defaultChannel, false)
+			return bundle.BuildFunc(path, "", bundleImage+":"+tag, containerTool, "", "", "", false)
```
Comment (kevinrizza) [inline test/e2e/opm_test.go:76]: Same comment here. Can we have a separate test for this rather than modifying an existing one?

File: pkg/lib/bundle/generate_test.go (hunk 4 above)
Comment (Bowenislandsong) [reply pkg/lib/bundle/generate_test.go:210]: right, I agree with the previous judgment about the e2e. but the change does alter the outcome for this specific unit test.

File: pkg/lib/bundle/generate.go (hunk 2 above)
Comment (Bowenislandsong) [reply pkg/lib/bundle/generate.go:15]: yeah, i see your point, just loved that library. getting rid of it.

File: pkg/lib/bundle/generate_test.go (hunk 4 above)
Comment (kevinrizza) [reply pkg/lib/bundle/generate_test.go:210]: Ah, this was actually invalid and we weren't catching it. Gotcha

File: pkg/lib/bundle/generate.go (hunk 3 above)
Comment (Bowenislandsong) [reply pkg/lib/bundle/generate.go:306]: ha, I copied that misspell and didn't even see it https://github.com/operator-framework/operator-registry/blob/a25687c90bbacce3c200a232376fa4952d22d307/pkg/lib/bundle/generate.go#L319

Comment (Bowenislandsong) [reply pkg/lib/bundle/generate.go:306]: https://play.golang.org/p/uemuPhlTisX found out that this split will always output an array with at least 1 length. My bad.

**Comment by dinhxuanvu** [comment]
Looking good.
/approve
Just a quick question. Why do we need the aqua bundle for this fix?

**Comment by Bowenislandsong** [comment]
> Looking good.
> /approve
> Just a quick question. Why do we need the aqua bundle for this fix?

Thanks, Kevin suggested to add a separate thing for e2e test that I am intereted and not to change existing package for my test purpose.

File: manifests/aqua/0.0.1/aquagateways.operator.aquasec.com.crd.yaml [hunk 6]
```diff
@@ -0,0 +1,34 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (kevinrizza) [inline manifests/aqua/0.0.1/aquagateways.operator.aquasec.com.crd.yaml:1]: Rather than add something to the root of the `manifests/` folder, can we just add these to a `test_data` folder for a separate test?

Comment (Bowenislandsong) [reply manifests/aqua/0.0.1/aquagateways.operator.aquasec.com.crd.yaml:1]: awww, I changed all those unit tests for it. Shouda thought about it... making the change now
